### ISSUE TITLE
On the implementation of pointwise convolution 

### ISSUE THREAD

**Issue by FanghaiZhao** [RElbers/ada-conv-pytorch#3]
Thanks for the authors' hard work of this code repository.

I have noticed that in current implementation, the pointwise convolution is also generated and viewed as group convolution (https://github.com/RElbers/ada-conv-pytorch/blob/5d0578ee8a8baf7ce06e129a933be2f2812145b3/lib/adaconv/kernel_predictor.py#L23). For example, with input channel C_in and output channel C_out, the generated pointwise convolution is with size C_out x C_in/N_g x 1 x 1. However, according my understandig I wonder if the pointwise convolution should have the same size as a normal convolution kernel with size C_out x C_in x 1 x1 ?

Looking forward to your respose. Thanks a lot.

**Comment by RElbers** [comment]
I see what you mean. In the Adaptive Convolutions paper they specify that the pointwise kernel is also grouped, but typically the pointwise convolution is a standard 1x1 convolution and not a grouped one. I'm not sure why the authors made that choice. It's been a while since I worked on this project, so I'll double check when I have some more time on my hands.

**Comment by FanghaiZhao** [comment]
Thank you very much for your response. Yes I noticed that in the paper the authors use the grouped version for the pointwise convolution. Not sure if it's too computational cost of generating parameters for a convolution kernel sized in C_out x C_in x 1 x 1 than its grouped version sized in C_out x C_in/N_g x 1 x 1. However in this way the pointwise convolution cannot play the role of interacting features across channels as in the original depthwise convolutiion.

**Comment by RElbers** [comment]
To give you a little update. Changing the 1x1 to normal convolutions doesn't change much. Training with the same learning rate didn't work though. The model almost immediately collapsed. But lowering the learning rate by a factor 10 fixed it. From what I could tell, the quality of the model was pretty much the same and the speed was also not impacted significantly. For both grouped and normal 1x1 convs it was around 1.2 iterations per second.

The model is quite sensitive to the learning rate in general I found, even with grouped convolutions. 
Because the model is creating the conv weights by transforming an input matrix vs. learning it via gradient descent, I think it just might be unstable in the beginning of training. As these transformations are bad when they are just initialized and with small batch-sizes the gradient is relatively noisy. I wonder if the adaptive learning rates of Adam might also be a problem. It's possible the model is more stable with SGD instead.
